framework/web/Application.php

```diff
@@ -55,7 +55,72 @@ class Application extends \yii\base\Application
      * @var Controller the currently active controller instance
      */
     public $controller;
+    /**
+     * Stack of middleware
+     * @var array
+     */
+    protected $middlewareStack = [];
 
+    /**
+     * Add middleware to stack
+     * @param $middleware callable|string|array
+     * Array format
+     * ```php
+     * [
+     *  'middleware' => 'className', //class name or closure
+     *  'priority' => 1, //integer value, middleware will be executed in desc order (default null)
+     *  'only' => ['index', 'view'], // middleware will be executed for only this actions (default for all actions)
+     *  'except' => ['update'], // middleware will not be executed for this actions (default empty array)
+     * ]
+     * ```
+     * @throws MiddlewareException
+     */
+    public function addMiddleware($middleware)
+    {
+        if (is_callable($middleware)) {
+            $middleware = [
+                'handle' => $middleware
+            ];
+        } elseif (is_string($middleware)) {
+            if (!class_exists($middleware)) {
```

Comment:
I'd remove this check as it will immediately trigger autoloading and may affect performance.